track/gps_client.py

```diff
@@ -0,0 +1,247 @@
+        BAD_TIME  = auto() # gps time != system time (suggests a ntpd<-->gpsd sync failure)
+        ERR_LAT   = auto() # excessive error in parameter: lat
+        ERR_LON   = auto() # excessive error in parameter: lon
+        ERR_ALT   = auto() # excessive error in parameter: alt
+        ERR_TRACK = auto() # excessive error in parameter: track
+        ERR_SPEED = auto() # excessive error in parameter: speed
+        ERR_CLIMB = auto() # excessive error in parameter: climb
+        ERR_TIME  = auto() # excessive error in parameter: time
+
+    # when you get this exception:
+    # - look at the flags set in the exception's reason field to see which things did not pass
+    # - look at the GPS object's fix_type, values, and errors fields to get exact information
+    class GetLocationFailure(Exception):
+        def __init__(self, reason):
+            super().__init__(str(reason))
+            self.reason = reason
+
+    def __init__(self, host='::1', port=gps.GPSD_PORT):
+        self.client = gps.gps(host, port, verbose=0, mode=gps.WATCH_ENABLE, reconnect=True)
+
+        self.fix_type = self.INIT_FIX
+        self.values   = self.INIT_VAL
+        self.errors   = self.INIT_ERR
+
+    def __enter__(self):
+        """Support usage of this class in 'with' statements."""
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        """Support usage of this class in 'with' statements."""
+        self.close()
+
+    def close(self):
+        self.client.close()
+
+    # this is the method that will be called by alignment code
+    # requirements for successful execution:
+    # - don't time out
+    # - get a 3D fix
+    # - all relevant errors need to be within the specified bounds
+    # - we don't care about DGPS
+    # on success: returns an astropy.coordinates.EarthLocation value
+    # on failure: raises GetLocationFailure with flags showing which requirements were not met
+    # (also may raise RuntimeError if something screwy happens with gpsd)
+    # parameters:
+    # - timeout: how much time to spend attempting to get a good fix before giving up
+    #            (set to inf for no timeout, i.e. try forever)
+    # - need_3d: whether a 3D fix should be considered necessary
+    # - err_max: a GPSValues tuple containing the thresholds that each 95% error value must be below
+    #            for the fix to be considered satisfactory
+    #            (individual members of the tuple may be set to inf to request no checking)
+    # - margins: a GPSMargins tuple indicating how far from zero the speed and climb values can be,
+    #            and how far from the system time the GPS time can be, for a satisfactory fix
+    #            (individual members of the tuple may be set to inf to request no checking)
+    # TODO: we *may* (not sure!) need to ensure that we get N consecutive reports that all meet our
+    #       requirements, before we declare ourselves successful and victorious
```

Comment:
I don't know whether this is necessary. As long as gpsd is good about not reporting stale information I'm not too worried about it. Omit for now.